File: src/client/xnative/lion.xdroid/MainActivity.cs
using Android.App;
using Android.Content.PM;
using Android.Gms.Ads;
using Android.Graphics;
using Android.OS;
using Android.Runtime;
using Android.Support.Design.Widget;
using Android.Support.V4.View;
using Android.Support.V7.App;
using Android.Views;
using Android.Widget;
using Firebase;
using Lion.XDroid.Control;
using Lion.XDroid.Libs;

namespace Lion.XDroid
{
    [Activity(Label = "MainActivity", Theme = "@style/DefaultTheme", ScreenOrientation = ScreenOrientation.Portrait, MainLauncher = true)]
    public class MainActivity : AppCompatActivity
    {
        MainPagerAdapter mainPagerAdapter;

        int prev_tab_position = 0;

        TabLayout main_tab_layout;
        ViewPager main_view_pager;

        Color selected_color = Color.ParseColor("#3296DC");
        Color unselect_color = Color.ParseColor("#9c9c9c");

        protected override void OnCreate(Bundle savedInstanceState)
        {
            base.OnCreate(savedInstanceState);
            Xamarin.Essentials.Platform.Init(this, savedInstanceState);
            // Set our view from the "main" layout resource
            SetContentView(Resource.Layout.activity_main);

            main_tab_layout = FindViewById<TabLayout>(Resource.Id.tl_main);
            main_view_pager = FindViewById<ViewPager>(Resource.Id.vp_main);

            ///----------------------------------------------------------------------------------------------------
            /// admob 오류시 참조 
            /// 
            /// google-servics.json 문제 fix
            /// https://forums.xamarin.com/discussion/96263/default-firebaseapp-is-not-initialized-in-this-process
            /// https://bugzilla.xamarin.com/show_bug.cgi?id=56108#c12
            /// https://blog.xamarin.com/lightweight-ads-for-android-apps/
            /// 
            /// com.google.android.gms 오류 fix
            /// proguard.cfg
            /// https://forums.xamarin.com/discussion/95107/firebaseinstanceidreceiver-classnotfoundexception-when-receiving-notifications
            /// https://developer.xamarin.com/guides/android/deployment,_testing,_and_metrics/proguard/#enabling
            ///----------------------------------------------------------------------------------------------------
            FirebaseApp.InitializeApp(this);

            this.InitTabLayout();
            this.InitEvent();
            this.InitAdMob();
        }

        public override void OnRequestPermissionsResult(int requestCode, string[] permissions, [GeneratedEnum] Android.Content.PM.Permission[] grantResults)
        {
            Xamarin.Essentials.Platform.OnRequestPermissionsResult(requestCode, permissions, grantResults);

            base.OnRequestPermissionsResult(requestCode, permissions, grantResults);
        }
        protected override void OnResume()
        {
            base.OnResume();

            if (this.mainPagerAdapter.IsAnonymous(prev_tab_position) == false)
            {
                var app_cache = new AppPreferences(this.ApplicationContext);
                var _token = app_cache.UserTokenKey;
                if (AppCommon.SNG.CheckExpired(this.ApplicationContext, _token) == false)
                {
                    var _fragment = (IPagerFragment)this.mainPagerAdapter.GetItem(prev_tab_position);
                    _fragment.Initialize();
                }
                else
                    prev_tab_position = 0;
            }

            var tab = main_tab_layout.GetTabAt(prev_tab_position);
            tab.Select();

            main_view_pager.CurrentItem = prev_tab_position;
        }

        private void InitAdMob()
        {
            MobileAds.Initialize(this.ApplicationContext, GetString(Resource.String.admob_app_id));

            var adView = FindViewById<AdView>(Resource.Id.adView);
            var adRequest = (new AdRequest.Builder()).Build();
            adView.LoadAd(adRequest);
        }

        private void TabColorToggle(int tab_position, bool selected)
        {
            var _color = selected_color;
            if (selected == false)
                _color = unselect_color;

            var _tab = main_tab_layout.GetTabAt(tab_position);
            {
                var _text = _tab.CustomView.FindViewById<TextView>(Resource.Id.tv_tab);
                _text.SetTextColor(_color);

                var _image = _tab.CustomView.FindViewById<ImageView>(Resource.Id.iv_tab);
                _image.SetColorFilter(_color, PorterDuff.Mode.SrcIn);
            }
        }

        private void InitEvent()
        {
            main_tab_layout.TabSelected += (sender, e) =>
            {
                TabColorToggle(prev_tab_position, false);

                //Log.Debug(this.GetType().Name, $"InitEvent ({prev_tab_position}) => ({e.Tab.Position})");
                prev_tab_position = e.Tab.Position;

                if (this.mainPagerAdapter.IsAnonymous(e.Tab.Position) == false)
                {
                    var app_cache = new AppPreferences(this.ApplicationContext);
                    var _token = app_cache.UserTokenKey;
                    if (AppCommon.SNG.CheckExpired(this.ApplicationContext, _token) == true)
                    {
                        AppCommon.SNG.JumpLoginActivity(this.ApplicationContext);
                        return;
                    }
                }

                main_view_pager.CurrentItem = e.Tab.Position;

                var _fragment = (IPagerFragment)this.mainPagerAdapter.GetItem(e.Tab.Position);
                _fragment.Initialize();

                TabColorToggle(e.Tab.Position, true);
            };
        }

        private void InitTabLayout()
        {
            this.mainPagerAdapter = new MainPagerAdapter(this.SupportFragmentManager);
            {
                var _winner = new ResultPagerFragment();
                {
                    this.mainPagerAdapter.AddFragment(_winner, "추첨결과", true);
                }

                var _number = new NumPagerFragment();
                {
                    this.mainPagerAdapter.AddFragment(_number, "번호관리", false);
                }

                var _store = new StorePagerFragment();
                {
                    this.mainPagerAdapter.AddFragment(_store, "로또판매점", true);
                }

                var _setting = new SettingPagerFragment();
                {
                    this.mainPagerAdapter.AddFragment(_setting, "설정", false);

                    _setting.Refresh += (s, e) =>
                    {
                        TabColorToggle(prev_tab_position, false);

                        this.OnResume();
                    };
                }
            }

            main_view_pager.Adapter = this.mainPagerAdapter;

            main_tab_layout.SetupWithViewPager(main_view_pager);

            this.SetupTabIcons(main_tab_layout);
        }

        private void SetupTabIcons(TabLayout tablayout)
        {
            View view0 = LayoutInflater.From(this).Inflate(Resource.Layout.main_tab, null);
            {
                var iv_tab0 = view0.FindViewById<ImageView>(Resource.Id.iv_tab);
                var tv_tab0 = view0.FindViewById<TextView>(Resource.Id.tv_tab);
                tv_tab0.SetText(this.mainPagerAdapter.GetPageTitle(0));
                iv_tab0.SetImageResource(Resource.Drawable.main1);
                tv_tab0.SetTextColor(this.selected_color);
                iv_tab0.SetColorFilter(this.selected_color, PorterDuff.Mode.SrcIn);
                tablayout.GetTabAt(0).SetCustomView(view0);
            }

            View view1 = LayoutInflater.From(this).Inflate(Resource.Layout.main_tab, null);
            {
                var iv_tab1 = view1.FindViewById<ImageView>(Resource.Id.iv_tab);
                var tv_tab1 = view1.FindViewById<TextView>(Resource.Id.tv_tab);
                tv_tab1.SetText(this.mainPagerAdapter.GetPageTitle(1));
                iv_tab1.SetImageResource(Resource.Drawable.main2);
                tv_tab1.SetTextColor(this.unselect_color);
                iv_tab1.SetColorFilter(this.unselect_color, PorterDuff.Mode.SrcIn);
                tablayout.GetTabAt(1).SetCustomView(view1);
            }

            View view2 = LayoutInflater.From(this).Inflate(Resource.Layout.main_tab, null);
            {
                var iv_tab2 = view2.FindViewById<ImageView>(Resource.Id.iv_tab);
                var tv_tab2 = view2.FindViewById<TextView>(Resource.Id.tv_tab);
                tv_tab2.SetText(this.mainPagerAdapter.GetPageTitle(2));
                iv_tab2.SetImageResource(Resource.Drawable.main3);
                tv_tab2.SetTextColor(this.unselect_color);
                iv_tab2.SetColorFilter(this.unselect_color, PorterDuff.Mode.SrcIn);
                tablayout.GetTabAt(2).SetCustomView(view2);
            }

            View view3 = LayoutInflater.From(this).Inflate(Resource.Layout.main_tab, null);
            {
                var iv_tab3 = view3.FindViewById<ImageView>(Resource.Id.iv_tab);
                var tv_tab3 = view3.FindViewById<TextView>(Resource.Id.tv_tab);
                tv_tab3.SetText(this.mainPagerAdapter.GetPageTitle(3));
                iv_tab3.SetImageResource(Resource.Drawable.main4);
                tv_tab3.SetTextColor(this.unselect_color);
                iv_tab3.SetColorFilter(this.unselect_color, PorterDuff.Mode.SrcIn);
                tablayout.GetTabAt(3).SetCustomView(view3);
            }
        }
    }
}

File: src/client/xcode/Lion.Droid/app/src/main/java/com/odinsoftware/lottolion/MainActivity.java
package com.odinsoftware.lottolion;

import android.content.Intent;
import android.graphics.Color;
import android.graphics.PorterDuff;
import android.media.Image;
import android.support.design.widget.TabLayout;
import android.support.v4.view.ViewPager;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.AdView;
import com.google.android.gms.ads.MobileAds;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.odinsoftware.lottolion.control.MainPagerAdapter;
import com.odinsoftware.lottolion.control.NumPagerFragment;
import com.odinsoftware.lottolion.control.ResultPagerFragment;
import com.odinsoftware.lottolion.control.SettingPagerFragment;
import com.odinsoftware.lottolion.control.StorePagerFragment;
import com.odinsoftware.lottolion.service.ApiAsnycTask;
import com.odinsoftware.lottolion.service.LLCommon;
import com.odinsoftware.lottolion.service.LLSharedPreferences;
import com.odinsoftware.lottolion.type.ApiResult;

import java.util.HashMap;

public class MainActivity extends AppCompatActivity {

    MainPagerAdapter mainPagerAdapter;
    int selectColor = Color.parseColor("#3296DC");
    int deSelectColor = Color.parseColor("#9c9c9c");
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        InitTab();
        InitEvent();
        InitAdMob();

    }

    private void InitAdMob() {
        MobileAds.initialize(getApplicationContext(), getString(R.string.banner_ad_unit_id));
        AdView mAdView = (AdView) findViewById(R.id.adView);
        AdRequest adRequest = new AdRequest.Builder().build();
        mAdView.loadAd(adRequest);
    }

    private void InitEvent() {
        TabLayout tl_main = (TabLayout)findViewById(R.id.tl_main);
        tl_main.setOnTabSelectedListener(listener_tab);
    }

    private void InitTab() {
        mainPagerAdapter = new MainPagerAdapter(getSupportFragmentManager());

        ResultPagerFragment result1 = new ResultPagerFragment();
        NumPagerFragment result2 = new NumPagerFragment();
        StorePagerFragment result3 = new StorePagerFragment();
        SettingPagerFragment result4 = new SettingPagerFragment ();
        mainPagerAdapter.addFragment(result1);
        mainPagerAdapter.addFragment(result2);
        mainPagerAdapter.addFragment(result3);
        mainPagerAdapter.addFragment(result4);
        result4.refresh=new SettingPagerFragment.Refresh() {
            @Override
            public void onRefresh() {
                onResume();
            }
        };

        ViewPager vp_main = (ViewPager) findViewById(R.id.vp_main);
        vp_main.setAdapter(mainPagerAdapter);

        TabLayout tl_main = (TabLayout)findViewById(R.id.tl_main);
        tl_main.setupWithViewPager(vp_main);
        setupTabIcons();
    }

    @Override
    protected void onResume() {
        super.onResume();
        TabLayout tl_main = (TabLayout)findViewById(R.id.tl_main);
        if(LLCommon.Get().CheckExpired(LLSharedPreferences.LL(getApplicationContext()).getSaveSession()) && (position == 1 || position == 3)) {
            TabLayout.Tab tab = tl_main.getTabAt(0);
            tab.select();
        }
        else{
            if(mainPagerAdapter.getItem(position) instanceof NumPagerFragment){
                TabLayout.Tab tab = tl_main.getTabAt(position);
                tab.select();
                ViewPager vp_main = (ViewPager) findViewById(R.id.vp_main);
                vp_main.setCurrentItem(position);
                ((NumPagerFragment)mainPagerAdapter.getItem(position)).Init();
            }
            if(mainPagerAdapter.getItem(position) instanceof SettingPagerFragment){
                TabLayout.Tab tab = tl_main.getTabAt(position);
                tab.select();
                ViewPager vp_main = (ViewPager) findViewById(R.id.vp_main);
                vp_main.setCurrentItem(position);
                ((SettingPagerFragment)mainPagerAdapter.getItem(position)).Init();
            }
        }
    }

    int position=0;
    TabLayout.OnTabSelectedListener listener_tab = new TabLayout.OnTabSelectedListener() {
        @Override
        public void onTabSelected(TabLayout.Tab tab) {
            position = tab.getPosition();
            ViewPager vp_main = (ViewPager) findViewById(R.id.vp_main);
            if(LLCommon.Get().CheckExpired(LLSharedPreferences.LL(getApplicationContext()).getSaveSession())&& (tab.getPosition() ==1 || tab.getPosition() ==3)){
                vp_main.setCurrentItem(0);
                Intent intent = new Intent(MainActivity.this, LoginActivity.class);
                startActivity(intent);
            }
            else{
                if (tab.getPosition() != vp_main.getCurrentItem()) {
                    vp_main.setCurrentItem(tab.getPosition());
                }
                if(mainPagerAdapter.getItem(position) instanceof NumPagerFragment){
                    ((NumPagerFragment)mainPagerAdapter.getItem(position)).Init();
                }
                if(mainPagerAdapter.getItem(position) instanceof SettingPagerFragment){
                    ((SettingPagerFragment)mainPagerAdapter.getItem(position)).Init();
                }

                TextView tv_tab = (TextView) tab.getCustomView().findViewById(R.id.tv_tab);
                tv_tab.setTextColor(selectColor);
                ImageView iv_tab = (ImageView) tab.getCustomView().findViewById(R.id.iv_tab);
                iv_tab.setColorFilter(selectColor,PorterDuff.Mode.SRC_IN);
            }
        }

        @Override
        public void onTabUnselected(TabLayout.Tab tab) {
            ViewPager vp_main = (ViewPager) findViewById(R.id.vp_main);
            if (tab.getPosition() != vp_main.getCurrentItem()) {
                vp_main.setCurrentItem(tab.getPosition());
            }
            TextView tv_tab = (TextView) tab.getCustomView().findViewById(R.id.tv_tab);
            tv_tab.setTextColor(deSelectColor);
            ImageView iv_tab = (ImageView) tab.getCustomView().findViewById(R.id.iv_tab);
            iv_tab.setColorFilter(deSelectColor,PorterDuff.Mode.SRC_IN);
        }

        @Override
        public void onTabReselected(TabLayout.Tab tab) {
            String s = "";
        }
    };
    private void setupTabIcons() {
        TabLayout tl_main = (TabLayout)findViewById(R.id.tl_main);
        View view1 = LayoutInflater.from(this).inflate(R.layout.main_tab, null);
        ImageView iv_tab1 = (ImageView)view1.findViewById(R.id.iv_tab);
        TextView tv_tab1 = (TextView)view1.findViewById(R.id.tv_tab);
        tv_tab1.setText(mainPagerAdapter.getPageTitle(0));
        iv_tab1.setImageResource(R.drawable.main1);
        tv_tab1.setTextColor(selectColor);
        iv_tab1.setColorFilter(selectColor,PorterDuff.Mode.SRC_IN);
        tl_main.getTabAt(0).setCustomView(view1);


        View view2 = LayoutInflater.from(this).inflate(R.layout.main_tab, null);
        ImageView iv_tab2 = (ImageView)view2.findViewById(R.id.iv_tab);
        TextView tv_tab2 = (TextView)view2.findViewById(R.id.tv_tab);
        tv_tab2.setText(mainPagerAdapter.getPageTitle(1));
        iv_tab2.setImageResource(R.drawable.main2);
        tv_tab2.setTextColor(deSelectColor);
        iv_tab2.setColorFilter(deSelectColor,PorterDuff.Mode.SRC_IN);
        tl_main.getTabAt(1).setCustomView(view2);

        View view3 = LayoutInflater.from(this).inflate(R.layout.main_tab, null);
        ImageView iv_tab3 = (ImageView)view3.findViewById(R.id.iv_tab);
        TextView tv_tab3 = (TextView)view3.findViewById(R.id.tv_tab);
        tv_tab3.setText(mainPagerAdapter.getPageTitle(2));
        iv_tab3.setImageResource(R.drawable.main3);
        tv_tab3.setTextColor(deSelectColor);
        iv_tab3.setColorFilter(deSelectColor,PorterDuff.Mode.SRC_IN);
        tl_main.getTabAt(2).setCustomView(view3);

        View view4 = LayoutInflater.from(this).inflate(R.layout.main_tab, null);
        ImageView iv_tab4 = (ImageView)view4.findViewById(R.id.iv_tab);
        TextView tv_tab4 = (TextView)view4.findViewById(R.id.tv_tab);
        tv_tab4.setText(mainPagerAdapter.getPageTitle(3));
        iv_tab4.setImageResource(R.drawable.main4);
        tv_tab4.setTextColor(deSelectColor);
        iv_tab4.setColorFilter(deSelectColor,PorterDuff.Mode.SRC_IN);
        tl_main.getTabAt(3).setCustomView(view4);
    }


}
